sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubUnboundedSink.java

```diff
@@ -0,0 +1,321 @@
+  private static final Logger LOG = LoggerFactory.getLogger(PubsubUnboundedSink.class);
+
+  /**
+   * Maximum number of messages per publish.
+   */
+  private static final int PUBLISH_BATCH_SIZE = 1000;
+
+  /**
+   * Maximum size of a publish batch, in bytes.
+   */
+  private static final long PUBLISH_BATCH_BYTES = 400000;
+
+  /**
+   * Longest delay between receiving a message and pushing it to pub/sub.
+   */
+  private static final Duration MAX_LATENCY = Duration.standardSeconds(2);
+
+  /**
+   * Period of samples for stats.
+   */
+  private static final Duration SAMPLE_PERIOD = Duration.standardMinutes(1);
+
+  /**
+   * Period of updates for stats.
+   */
+  private static final Duration SAMPLE_UPDATE = Duration.standardSeconds(5);
+
+  /**
+   * How frequently to log stats.
+   */
+  private static final Duration LOG_PERIOD = Duration.standardSeconds(30);
+
+  /**
+   * Additional sharding so that we can hide publish message latency.
+   */
+  private static final int SCALE_OUT = 4;
+
+  public static final Coder<PubsubClient.OutgoingMessage> CODER = new
+      AtomicCoder<PubsubClient.OutgoingMessage>() {
+        @Override
+        public void encode(
+            PubsubClient.OutgoingMessage value, OutputStream outStream, Context context)
+            throws CoderException, IOException {
+          ByteArrayCoder.of().encode(value.elementBytes, outStream, Context.NESTED);
+          BigEndianLongCoder.of().encode(value.timestampMsSinceEpoch, outStream, Context.NESTED);
+        }
+
+        @Override
+        public PubsubClient.OutgoingMessage decode(
+            InputStream inStream, Context context) throws CoderException, IOException {
+          byte[] elementBytes = ByteArrayCoder.of().decode(inStream, Context.NESTED);
+          long timestampMsSinceEpoch = BigEndianLongCoder.of().decode(inStream, Context.NESTED);
+          return new PubsubClient.OutgoingMessage(elementBytes, timestampMsSinceEpoch);
+        }
+      };
+
+  // ================================================================================
+  // ShardFv
+  // ================================================================================
+
+  /**
+   * Convert elements to messages and shard them.
+   */
+  private class ShardFn extends DoFn<T, KV<Integer, PubsubClient.OutgoingMessage>> {
+    /**
+     * Number of cores available for publishing.
+     */
+    private final int numCores;
+
+    private final Aggregator<Long, Long> elementCounter =
+        createAggregator("elements", new Sum.SumLongFn());
+
+    public ShardFn(int numCores) {
+      this.numCores = numCores;
+    }
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      elementCounter.addValue(1L);
+      byte[] elementBytes = CoderUtils.encodeToByteArray(elementCoder, c.element());
+      long timestampMsSinceEpoch = c.timestamp().getMillis();
+      c.output(KV.of(ThreadLocalRandom.current().nextInt(numCores * SCALE_OUT),
```

Comment:
Using a random int will allow most systems to choose the sharding to an even more arbitrary degree, and then you can remove numCores